### PR TITLE
Add Separator and SeparatorEditor

### PR DESCRIPTION
**Pull request by jentao**
Screen Shot 2019-07-15 at 12 32 31 AM

I replaced all sliders with numeric input boxes. I realized that sliding effect makes the webpage rerender a lot and using input boxes might be better.

### CODE REVIEW

File: src/components/ContentMapping/ContentMapping.ts [hunk 1]
```diff
@@ -5,6 +5,8 @@ import { Stub, StubEditor } from '../ContentWidgets/_Stub/Stub';
 import { ContentSingularData } from '../_data/ContentSingularData';
 import { ExampleImage } from '../ContentWidgets/ExampleImage/ExampleImage';
 import { ExampleImageEditor } from '../ContentWidgets/ExampleImage/ExampleImageEditor';
+import { Separator } from "../ContentWidgets/Separator/Separator";
```
Comment (kwokwilliam) [inline src/components/ContentMapping/ContentMapping.ts]: Nitpick: Please use consistency for imports-- use single quotations here.

File: src/components/ContentMapping/ContentMapping.ts [hunk 2]
```diff
@@ -37,5 +40,10 @@ export const ContentMapping: ContentMappingType = {
     EXAMPLE_IMAGE: {
         widget: ExampleImage,
         editor: ExampleImageEditor
-    }
-}
\ No newline at end of file
+    },
+    SEPARATOR: {
+    widget: Separator,
```
Comment (kwokwilliam) [inline src/components/ContentMapping/ContentMapping.ts]: Nitpick: Tab this out

File: src/components/_data/ContentSingularData.ts [hunk 3]
```diff
@@ -1,6 +1,8 @@
 export type ContentSingularData = {
-    plainText_content?: string
-    exampleImage_imageLink?: string
-    exampleImage_percentageSize?: number
-    [idx: string]: any
-}
\ No newline at end of file
+  plainText_content?: string;
```
Comment (kwokwilliam) [inline src/components/_data/ContentSingularData.ts]: Nitpick: Tab this out as well

File: src/components/ContentWidgets/Separator/SeparatorEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+ *
+ * Last Modified
+ * Jennifer Tao
+ * July 10, 2019
+ */
+export const SeparatorEditor: React.FC<WidgetEditorProps> = ({
+  editedContent,
+  setEditedContentOnChange
+}) => {
+  const defaultColor = { r: 66, g: 13, b: 171 };
```
Comment (kwokwilliam) [inline src/components/ContentWidgets/Separator/SeparatorEditor.tsx]: Declare this const in a constants file instead, otherwise you will create a new version of this when there are multiple editors for a Separator. Small optimizations.

File: src/components/ContentWidgets/Separator/SeparatorEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  const width = editedContent.separator_width || 33;
```
Comment (kwokwilliam) [inline src/components/ContentWidgets/Separator/SeparatorEditor.tsx]: Set all these values, `33`, `"red"`, `"green"`, `"blue"` out into constants defined either above outside of this function or in some sort of constants file. That way we aren't reliant on just strings.

Actually, for color I would do like an enum or something instead.

```tsx
enum Color {
  RED = "RED", 
  GREEN = "GREEN", 
  BLUE = "BLUE"
}

File: src/components/ContentWidgets/Separator/SeparatorEditor.tsx [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+  /**
+   * Updates seperator_color
+   * @param e event
+   * @param c color: {red, green, blue}
+   */
+  let changeColor = (e: any, c: string) => {
+    let newColor = { ...color };
+    if (c === "red") {
+      newColor.r = e.target.value;
```
Comment (kwokwilliam) [inline src/components/ContentWidgets/Separator/SeparatorEditor.tsx]: I would use a switch case here instead, and use that enum we defined earlier

File: src/components/ContentWidgets/Separator/SeparatorEditor.tsx [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+  const defaultColor = { r: 66, g: 13, b: 171 };
+  const width = editedContent.separator_width || 33;
+  const color = editedContent.separator_color || defaultColor;
+
+  /**
+   * Updates seperator_color
+   * @param e event
+   * @param c color: {red, green, blue}
+   */
+  let changeColor = (e: any, c: string) => {
```
Comment (kwokwilliam) [inline src/components/ContentWidgets/Separator/SeparatorEditor.tsx]: const instead of let here.

File: src/components/ContentWidgets/Separator/SeparatorEditor.tsx [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+      <Separator separator_width={width} separator_color={color} />
+      <h3>Size:</h3>
+      Width:
+      <input
+        className="slider width-range"
+        type="range"
+        min="0"
+        max="100"
+        defaultValue={width + ""}
+        onChange={e =>
```
Comment (kwokwilliam) [inline src/components/ContentWidgets/Separator/SeparatorEditor.tsx]: Keep in mind that this may be taxing on the computation side because you would be re-rendering every single tick the user changes the color. First, I would double check that it only calls `onChange` once the user lets go of their mouse. If that is the case, you are good here

If not, I would look into `onMouseUp` or something like that so it only updates once the user has let go.

File: src/components/ContentWidgets/Separator/SeparatorEditor.tsx [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    <>
+      <Separator separator_width={width} separator_color={color} />
+      <h3>Size:</h3>
+      Width:
+      <input
+        className="slider width-range"
+        type="range"
+        min="0"
+        max="100"
+        defaultValue={width + ""}
```
Comment (kwokwilliam) [inline src/components/ContentWidgets/Separator/SeparatorEditor.tsx]: You should instead do `string(width)` or something to typecast instead of using javascript's way of type casting.

Comment (kwokwilliam) [reply src/components/ContentWidgets/Separator/SeparatorEditor.tsx]: Also I don't think you should ever use `defaultValue`, instead, use `value` and keep that `value` equal to the `editedContent` value. This is what using a controlled form is all about.

File: src/components/ContentWidgets/Separator/SeparatorEditor.tsx [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+      {width}% (Default 33%)
+      <h3>Color:</h3>
+      <input
+        type="button"
+        value="Default Color"
+        onClick={() => {
+          setEditedContentOnChange("separator_color", defaultColor);
+        }}
+      />
+      Default: R: 66, G: 13, B: 171
```
Comment (kwokwilliam) [inline src/components/ContentWidgets/Separator/SeparatorEditor.tsx]: I think this is bad user experience, and instead we should have some sort of button that sets the default values or have some presets the user can choose from.

File: src/components/ContentWidgets/Separator/SeparatorEditor.tsx [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+      <div>
+        <input
+          className="slider red-range"
+          type="range"
+          min="0"
+          max="255"
+          defaultValue="66"
```
Comment (kwokwilliam) [inline src/components/ContentWidgets/Separator/SeparatorEditor.tsx]: no need for defaultValue, as far as I know.

File: src/components/ContentWidgets/Separator/SeparatorEditor.tsx [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        <input
+          className="slider green-range"
+          type="range"
+          min="0"
+          max="255"
+          defaultValue="13"
+          value={color.g}
+          onChange={e => changeColor(e, "green")}
+        />
+        Green:{" "}
```
Comment (kwokwilliam) [inline src/components/ContentWidgets/Separator/SeparatorEditor.tsx]: I would use `&nbsp;` instead of `{" "}`

File: src/components/ContentWidgets/Separator/Separator.tsx [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+ * July 10, 2019
+ */
+export const Separator: React.FC<ContentSingularData> = ({
+  separator_width,
+  separator_color
+}) => {
+  if (!separator_width || !separator_color) {
+    return <></>;
+  }
+  const color =
```
Comment (kwokwilliam) [inline src/components/ContentWidgets/Separator/Separator.tsx]: you can shorten this to 

```tsx
const {r, g, b} = separator_color;
const color = `rgb(${r},${g},${b})`;
```

File: src/components/ContentWidgets/Separator/Separator.tsx [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    "rgb(" +
+    separator_color.r +
+    "," +
+    separator_color.g +
+    "," +
+    separator_color.b +
+    ")";
+  return (
```
Comment (kwokwilliam) [inline src/components/ContentWidgets/Separator/Separator.tsx]: remove parenthesis

File: src/components/ContentMapping/ContentMapping.ts (hunk 2 above)
Comment (kwokwilliam) [reply src/components/ContentMapping/ContentMapping.ts]: Please set your tab width to 4 spaces for now for the entire repo (can be found in vscode settings). I will update this to regular tabs in the future so you can define how much spaces you want your tabs.

File: src/components/ContentWidgets/Separator/Separator.css [hunk 15]
```diff
@@ -0,0 +1,56 @@
+.separator {
```
Comment (kwokwilliam) [inline src/components/ContentWidgets/Separator/Separator.css]: Combine some of your css selectors by using commas. For example you can merge together something like this:

```css
.width-range::-webkit-slider-thumb, .width-range::-moz-range-thumb {
  background: black;
}
```

File: src/components/ContentWidgets/Separator/Separator.tsx [hunk 16]
```diff
@@ -0,0 +1,38 @@
+import React from "react";
```
Comment (kwokwilliam) [inline src/components/ContentWidgets/Separator/Separator.tsx]: nitpick: single quotations

**Review by kwokwilliam — CHANGES_REQUESTED** [review]
Otherwise looks good! Make these changes, and also include a picture of what it looks like with some settings, so we can see it without opening it in the browser.

File: src/components/ContentWidgets/Separator/Separator.css (hunk 15 above)
Comment (jentao) [reply src/components/ContentWidgets/Separator/Separator.css]: It seems like combining vendor-specific pseudo-elements will confuse the browser and ignore the css block. The slider handle will disappear like this:
Screen Shot 2019-07-14 at 9 10 54 PM

Comment (kwokwilliam) [reply src/components/ContentWidgets/Separator/Separator.css]: Interesting. No problem then, feel free to ignore

File: src/components/ContentWidgets/Separator/SeparatorEditor.tsx [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+import { Separator } from './Separator';
+import './SeparatorConstants';
+import { DEFAULT_WIDTH, DEFAULT_COLOR, Color } from './SeparatorConstants';
+
+/**
+ * SeparatorEditor edits the width and color of a separator widget
+ *
+ * Last Modified
+ * Jennifer Tao
+<<<<<<< HEAD
```
Comment (kwokwilliam) [inline src/components/ContentWidgets/Separator/SeparatorEditor.tsx]: unresolved merge conflict?

**Review by kwokwilliam — CHANGES_REQUESTED** [review]

**Comment by kwokwilliam** [comment]
I am starting to merge these pull requests, starting with the oldest.
